packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,203 @@
+///
+/// @author Brecht Devos - <brecht@loopring.org>
+contract FastWithdrawalAgent is ReentrancyGuard
+{
+    using AddressUtil       for address;
+    using AddressUtil       for address payable;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+    using SignatureUtil     for bytes32;
+
+    struct FastWithdrawal
+    {
+        address exchange;
+        address from;
+        address to;
+        address token;
+        uint    amount;
+        address feeToken;
+        uint    fee;
+        uint    validUntil;
+        bool    onchainFeePayment;
+        bytes   signature;
+    }
+
+    // Replay protection
+    mapping (bytes32 => bool) txHash;
+
+    // EIP712
+    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
+        "FastWithdrawal(address exchange,address from,address to,address token,uint256 amount,address feeToken,uint256 fee,bool onchainFeePayment,uint256 validUntil)"
+    );
+    bytes32 public DOMAIN_SEPARATOR;
+
+    constructor()
+        public
+    {
+        //DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+    }
+
+    function executeFastWithdrawals(FastWithdrawal[] memory fastWithdrawals)
+        public
+        payable
+        nonReentrant
+    {
+        // Do all fast withdrawals
+        for (uint i = 0; i < fastWithdrawals.length; i++) {
+            executeFastWithdrawal(fastWithdrawals[i]);
+        }
+        // Return any ETH left into this contract
+        // (can happen when more ETH is sent than needed for the fast withdrawals)
+        msg.sender.sendETHAndVerify(address(this).balance, gasleft());
+    }
+
+    // -- Internal --
+
+    function executeFastWithdrawal(FastWithdrawal memory fastWithdrawal)
+        internal
+    {
+        // Currently here because the compiler hangs when doing this in the constructor
+        if (DOMAIN_SEPARATOR == bytes32(0)) {
+            DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+        }
+
+        // Compute the hash
+        bytes32 hash = EIP712.hashPacked(
+            DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encodePacked(
+                    FASTWITHDRAWAL_TYPEHASH,
+                    fastWithdrawal.exchange,
+                    fastWithdrawal.from,
+                    fastWithdrawal.to,
+                    fastWithdrawal.token,
+                    fastWithdrawal.amount,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fee,
+                    fastWithdrawal.onchainFeePayment,
+                    fastWithdrawal.validUntil
+                )
+            )
+        );
+
+        // Make sure the tx isn't used multiple times
+        require(txHash[hash] == false, "TX_REPLAY");
+        txHash[hash] = true;
+
+        // Check the signature
+        require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
```

Comment:
in our transfer circuits, do we allow account to transfer to itself? If not, we need to check here that `fastWithdrawalFrom !=msg.sender`

Reply:
The circuits allow this (and this is tested).